src/cocotb/types/logic_array.py

```diff
@@ -391,15 +391,16 @@ def from_signed(
         """
         if isinstance(range, int):
             range = Range(range - 1, "downto", 0)
-        if value < 0:
-            value += 2 ** len(range)
-        # If value doesn't fit in range, it will still be negative and will blow the
-        # constructor up in a bad way.
-        if value < 0:
+        elif not isinstance(range, Range):
+            raise TypeError("Invalid range type")
+
+        limit = 1 << (len(range) - 1)
+        if (value < -limit) or (value >= limit):
```

Comment:
Unnecessary parens and prefer `<` when doing these bounds based checks so they can be easily read left-to-right, like the inverse of `-limit <= value < limit` where the limits are on the inside because the value lies outside.
```suggestion
        if value < -limit or limit <= value:
```